### ISSUE TITLE
Energy minimization with fix rigid/small

### ISSUE THREAD

**Issue by demkowicz** [lammps/lammps#3801]
Under the "Restrictions" heading on the bottom of the minimize command manual entry (https://docs.lammps.org/minimize.html), there is a statement you'd be interested in implementing energy minimization under the fix rigid constraint. I believe this could be achieved pretty easily, at least for fix rigid/small using derivative methods such as conjugate gradient. I'm not familiar with LAMMPS source code, but, if there is interest, I would be happy to work with someone who knows it to implement this functionality. I would personally very much like to see such functionality in LAMMPS.
M. J. Demkowicz (demkowicz@tamu.edu)

**Comment by athomps** [comment]
I could certainly give you some guidance.  I suggest starting out by looking at two functions: `MinLineSearch::alpha_step()` and `Min::energy_force()`.  Let me know how you feel after that :-)

**Comment by sjplimp** [comment]
You should look at this paper for HOOMD.  I don't have a copy at the moment, but it has a section on energy min with rigid body constraints.  My recollection is it uses FIRE (damped dynamics) not a CG based method.
@ndtrung81  Trung can comment as he is the lead author.  I think this is likely a better match to how LAMMPS would use MPI (or GPU) parallelism.

Nguyen, T. D., Phillips, C. L., Anderson, J. A., & Glotzer, S. C. (2011). Rigid body constraints realized in massively-parallel molecular dynamics on graphics processing units. Computer Physics Communications, 182(11), 2307-2313. https://doi.org/10.1016/j.cpc.2011.06.005

**Comment by ndtrung81** [comment]
Correct, in that paper we extended the FIRE algorithm (E. Bitzek et al, Phys. Rev. Lett. 97 (2006)) to account for the contribution of the body rotational degrees of freedom to the power (body torque times angular velocity in the space-fixed frame). I think it would be possible to either extend the current MinFire to handle rigid bodies (when fix rigid is present), or to have a separate MinFireRigid deriving from MinFire. The former approach would allow energy minimization of systems composed of non-rigid parts and rigid bodies but need changes to MinFire. In the latter, it is difficult to ensure two MinFire* instances update the atoms in the non-rigid parts and those in the bodies consistently.

**Comment by demkowicz** [comment]
Aidan,

I had a look, but it doesn’t really tell my anything. I’ve written my own implementations of CG minimization, so I understand the algorithm. However, I’ve never worked with LAMMPS source code.


—
Michael J Demkowicz, Professor
Texas A&M Materials Science and Engineering
***@***.******@***.***> / 979-458-9845 / RDMC 212

On May 31, 2023, at 12:20 PM, Aidan Thompson ***@***.******@***.***>> wrote:

I could certainly give you some guidance. I suggest starting out by looking at two functions: MinLineSearch: : alpha_step() and Min: : energy_force(). Let me know how you feel after that :-) — Reply to this email directly, view it on GitHub,
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender
This message came from outside your organization.

ZjQcmQRYFpfptBannerEnd


I could certainly give you some guidance. I suggest starting out by looking at two functions: MinLineSearch::alpha_step() and Min::energy_force(). Let me know how you feel after that :-)

**Comment by demkowicz** [comment]
I’m not familiar with the FIRE algorithm. However, overdamped dynamics is very close to constant increment steepest descent. Glancing over the paper, I see it has standard rigid body dynamics with the only apparent novelty being in the integration scheme. This makes me think it may be possible to do a quick kluge within lammps script where you run 1 tilmestep at a time in a loop. On each step, you set the COM velocity to the COM force and the angular velocity to the torque for each molecule. Then you adjust the tilmestep to limit the max displacement and rotation. This may be a relatively simple thing to try first with no need to alter the source code.

—
Michael J Demkowicz, Professor
Texas A&M Materials Science and Engineering
***@***.******@***.***> / 979-458-9845 / RDMC 212

On May 31, 2023, at 4:07 PM, Steve Plimpton ***@***.******@***.***>> wrote:

This Message Is From an External Sender
This message came from outside your organization.


You should look at this paper for HOOMD. I don't have a copy at the moment, but it has a section on energy min with rigid body constraints. My recollection is it uses FIRE (damped dynamics) not a CG based method.
@ndtrung81<https://urldefense.com/v3/__https://github.com/ndtrung81__;!!KwNVnqRv!CUqmCFWEN700XuIkdt53HJwi1-N1WNo4DKtPVkPmW0y9aXA4aXisWOJwrUYNPrk_sxh6Ioow5RT9VGW0iDfcY-bfDCg$> Trung can comment as he is the lead author. I think this is likely a better match to how LAMMPS would use MPI (or GPU) parallelism.

Nguyen, T. D., Phillips, C. L., Anderson, J. A., & Glotzer, S. C. (2011). Rigid body constraints realized in massively-parallel molecular dynamics on graphics processing units. Computer Physics Communications, 182(11), 2307-2313. https://doi.org/10.1016/j.cpc.2011.06.005<https://urldefense.com/v3/__https://doi.org/10.1016/j.cpc.2011.06.005__;!!KwNVnqRv!CUqmCFWEN700XuIkdt53HJwi1-N1WNo4DKtPVkPmW0y9aXA4aXisWOJwrUYNPrk_sxh6Ioow5RT9VGW0iDfcpxsTSGI$>

**Comment by demkowicz** [comment]
Trung, perhaps you would be interested in implementing this in LAMMPS? As far as I know, LAMMPS currently does not have any capability of doing energy minimization with rigid body constraints.

—
Michael J Demkowicz, Professor
Texas A&M Materials Science and Engineering
***@***.******@***.***> / 979-458-9845 / RDMC 212

On Jun 1, 2023, at 10:47 AM, Trung Nguyen ***@***.******@***.***>> wrote:

This Message Is From an External Sender
This message came from outside your organization.


Correct, in that paper we extended the FIRE algorithm (E. Bitzek et al, Phys. Rev. Lett. 97 (2006)) to account for the contribution of the body rotational degrees of freedom to the power (body torque times angular velocity in the space-fixed frame). I think it would be possible to either extend the current MinFire to handle rigid bodies (when fix rigid is present), or to have a separate MinFireRigid deriving from MinFire. The former approach would allow energy minimization of systems composed of non-rigid parts and rigid bodies but need changes to MinFire. In the latter, it is difficult to ensure two MinFire* instances update the atoms in the non-rigid parts and those in the bodies consistently.

**Comment by demkowicz** [comment]
BTW, Steve, why do you anticipate parallelization issues for conjugate gradient? Is it because the line search might lead to large displacements that could require remaking the linked list? This is possible in my experience, but not likely. The performance advantage of CG over steepest descent is rather, uh… steep! ;)

—
Michael J Demkowicz, Professor
Texas A&M Materials Science and Engineering
***@***.******@***.***> / 979-458-9845 / RDMC 212

On Jun 1, 2023, at 4:35 PM, Demkowicz, Michael J ***@***.******@***.***>> wrote:

I’m not familiar with the FIRE algorithm. However, overdamped dynamics is very close to constant increment steepest descent. Glancing over the paper, I see it has standard rigid body dynamics with the only apparent novelty being in the integration scheme. This makes me think it may be possible to do a quick kluge within lammps script where you run 1 tilmestep at a time in a loop. On each step, you set the COM velocity to the COM force and the angular velocity to the torque for each molecule. Then you adjust the tilmestep to limit the max displacement and rotation. This may be a relatively simple thing to try first with no need to alter the source code.

—
Michael J Demkowicz, Professor
Texas A&M Materials Science and Engineering
***@***.******@***.***> / 979-458-9845 / RDMC 212

On May 31, 2023, at 4:07 PM, Steve Plimpton ***@***.******@***.***>> wrote:

This Message Is From an External Sender
This message came from outside your organization.


You should look at this paper for HOOMD. I don't have a copy at the moment, but it has a section on energy min with rigid body constraints. My recollection is it uses FIRE (damped dynamics) not a CG based method.
@ndtrung81<https://urldefense.com/v3/__https://github.com/ndtrung81__;!!KwNVnqRv!CUqmCFWEN700XuIkdt53HJwi1-N1WNo4DKtPVkPmW0y9aXA4aXisWOJwrUYNPrk_sxh6Ioow5RT9VGW0iDfcY-bfDCg$> Trung can comment as he is the lead author. I think this is likely a better match to how LAMMPS would use MPI (or GPU) parallelism.

Nguyen, T. D., Phillips, C. L., Anderson, J. A., & Glotzer, S. C. (2011). Rigid body constraints realized in massively-parallel molecular dynamics on graphics processing units. Computer Physics Communications, 182(11), 2307-2313. https://doi.org/10.1016/j.cpc.2011.06.005<https://urldefense.com/v3/__https://doi.org/10.1016/j.cpc.2011.06.005__;!!KwNVnqRv!CUqmCFWEN700XuIkdt53HJwi1-N1WNo4DKtPVkPmW0y9aXA4aXisWOJwrUYNPrk_sxh6Ioow5RT9VGW0iDfcpxsTSGI$>

**Comment by athomps** [comment]
@demkowicz A lot of the difficulties with minimization in LAMMPS is that atoms *may* migrate between processors after every energy evaluation. This requires carefully updating all data structures, including those for rigid constraints. It is perfectly doable, but requires some effort.

**Comment by demkowicz** [comment]
Gotcha. But doesn’t this come up with regular dynamics, too?


—
Michael J Demkowicz, Professor
Texas A&M Materials Science and Engineering
***@***.*** / 979-458-9845 / RDMC 212

On Jun 1, 2023, at 4:50 PM, Aidan Thompson ***@***.***> wrote:

﻿
@demkowicz A lot of the difficulties with minimization in LAMMPS is that atoms may migrate between processors after every energy evaluation. This requires carefully updating all data structures, including those for rigid constraints. It is
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender
This message came from outside your organization.

ZjQcmQRYFpfptBannerEnd

@demkowicz<https://urldefense.com/v3/__https://github.com/demkowicz__;!!KwNVnqRv!HwHowefE5yRQTv2rVsi4tqSrFQTW3T_SYUyJm0Q7_d88GMKq6G6Uz7KH2taEsV0wP6RCgb7t362KabPoY6_5oTtHSjE$> A lot of the difficulties with minimization in LAMMPS is that atoms may migrate between processors after every energy evaluation. This requires carefully updating all data structures, including those for rigid constraints. It is perfectly doable, but requires some effort.

**Comment by athomps** [comment]
Yes, but not every timestep, because the halo region is buffered.  Also, there is something else. The line search methods, like CG, do backtracking, so they start out with a big step and then shrink it.

**Comment by demkowicz** [comment]
Yeah, I figured it would be something like that. Thanks for clarifying!

—
Michael J Demkowicz, Professor
Texas A&M Materials Science and Engineering
***@***.******@***.***> / 979-458-9845 / RDMC 212

On Jun 1, 2023, at 4:57 PM, Aidan Thompson ***@***.******@***.***>> wrote:

This Message Is From an External Sender
This message came from outside your organization.


Yes, but not every timestep, because the halo region is buffered. Also, there is something else. The line search methods, like CG, do backtracking, so they start out with a big step and then shrink it.

**Comment by demkowicz** [comment]
Well, I wrote my little poor man's rigid body relaxation algorithm. It works! It takes about 70 iterations (about 1s on a single CPU) to relax 64 tip4p H2O molecules in an ice Ih structure to within about 0.01eV. Script available upon request.

**Comment by mm5986** [comment]
> Well, I wrote my little poor man's rigid body relaxation algorithm. It works! It takes about 70 iterations (about 1s on a single CPU) to relax 64 tip4p H2O molecules in an ice Ih structure to within about 0.01eV. Script available upon request.

This is great! I've been looking for a rigid body relaxation algorithm. Could you please send/post your script.

**Comment by demkowicz** [comment]
See attached. Note that I have not done anything more than a proof of concept. In particular, I did not do any validation or optimization, so I give absolutely no guarantees on performance. There is also clear room for improvement, e.g., replacing the energy-based convergence criterion with a force/torque-based one.

[in.lammps_relax.txt](https://github.com/lammps/lammps/files/11964976/in.lammps_relax.txt)

**Comment by mm5986** [comment]
Thanks a ton. This will be so helpful.

**Comment by akohlmey** [comment]
@demkowicz, @ndtrung81 this has been sitting idle for over two years now. Are there any updates?

If not, I should test the attached script and add what is sufficient to explain the approach to the documentation.

**Comment by ndtrung81** [comment]
@akohlmey sorry I haven't done anything related to this.

**Comment by demkowicz** [comment]
Axel,

In short—I have not worked on it further.

At the time of this comment, I was doing molecular simulations using rigid body dynamics. Shortly afterwards, I switched to constraint dynamics because the usual minimizers run with it. However, I noticed that they are not very effective. In fact, I’ve been doing most of my relaxations using low temperature md. That is not ideal, especially since it is no help with neb. Overall, I do believe the topic deserves more attention.


—
Michael J Demkowicz, Professor
Texas A&M Materials Science and Engineering
***@***.*** / 979-458-9845 / RDMC 212

On Oct 20, 2025, at 7:14 AM, Axel Kohlmeyer ***@***.***> wrote:

﻿
akohlmey left a comment (lammps/lammps#3801) @demkowicz, @ndtrung81 this has been sitting idle for over two years now. Are there any updates? If not, I should test the attached script and add what is sufficient to explain the approach to the
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender
This message came from outside your organization.

ZjQcmQRYFpfptBannerEnd
[https://avatars.githubusercontent.com/u/289251?s=20&v=4]akohlmey left a comment (lammps/lammps#3801)<https://urldefense.com/v3/__https://github.com/lammps/lammps/issues/3801*issuecomment-3421815618__;Iw!!KwNVnqRv!HGuXFaqVTtuYpoFdfWveTVYpC_YpmOfZRggpEs1DYXLdNIPzl5lQZ0CO6tZ2UC5MpptSyisVBdBlZoACJWxxjWw047Q$>

@demkowicz<https://urldefense.com/v3/__https://github.com/demkowicz__;!!KwNVnqRv!HGuXFaqVTtuYpoFdfWveTVYpC_YpmOfZRggpEs1DYXLdNIPzl5lQZ0CO6tZ2UC5MpptSyisVBdBlZoACJWxxrEQn7Tk$>, @ndtrung81<https://urldefense.com/v3/__https://github.com/ndtrung81__;!!KwNVnqRv!HGuXFaqVTtuYpoFdfWveTVYpC_YpmOfZRggpEs1DYXLdNIPzl5lQZ0CO6tZ2UC5MpptSyisVBdBlZoACJWxxUN_rsqY$> this has been sitting idle for over two years now. Are there any updates?

If not, I should test the attached script and add what is sufficient to explain the approach to the documentation.

**Comment by akohlmey** [comment]
> However, I noticed that they are not very effective.

There was some unexpected issue with the fix shake restraint implementation used during minimization that was corrected only in the 10 Sep 2025 feature release and update 1 to the 22 July 2025 stable release. That allowed to increase the restraint forces by two orders of magnitude while reducing the number of steps required and no more divergence when the neighbor lists were updated.

> Overall, I do believe the topic deserves more attention.

There are lots of things that are *deserving*, but unfortunately, we have only limited workforce available. Most features and contributions in LAMMPS are made by external contributors. The core LAMMPS developers focus more on maintenance and specific tasks they are funded to do.

The topic to hire one of more research software engineers for the expressed purpose to implement requested LAMMPS features comes up repeatedly, but we have not yet agreed on a way to procure the required funding. Basically, there are four main paths: 1. obtain a grant from one or more federal or private funding agencies, 2. donations from users (especially commercial and institutional users), 3. crowdfunding, 4. create or join a non-profit organization created for the purpose of supporting Open Source software development. So far, it has not been obvious which is the best choice and how high the chances are for that choice to be successful.

**Comment by demkowicz** [comment]
Thank you, I’ll give a go to the new release at the next opportunity.

—
Michael J Demkowicz, Professor
Texas A&M Materials Science and Engineering
***@***.*** / 979-458-9845 / RDMC 212

On Oct 20, 2025, at 7:56 AM, Axel Kohlmeyer ***@***.***> wrote:

﻿
akohlmey left a comment (lammps/lammps#3801) However, I noticed that they are not very effective. There was some unexpected issue with the fix shake restraint implementation used during minimization that was corrected only in the 10 Sep 2025
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender
This message came from outside your organization.

ZjQcmQRYFpfptBannerEnd
[https://avatars.githubusercontent.com/u/289251?s=20&v=4]akohlmey left a comment (lammps/lammps#3801)<https://urldefense.com/v3/__https://github.com/lammps/lammps/issues/3801*issuecomment-3421950358__;Iw!!KwNVnqRv!FB5rVZjULxAGwOeOBGjjfBN0Iv5_xVxBo8_SiriXzrqHLpdQPzydoH1XaBVI-jvckIDuX83rtVsRrgFm2_wdtEbil54$>

However, I noticed that they are not very effective.

There was some unexpected issue with the fix shake restraint implementation used during minimization that was corrected only in the 10 Sep 2025 feature release and update 1 to the 22 July 2025 stable release. That allowed to increase the restraint forces by two orders of magnitude while reducing the number of steps required and no more divergence when the neighbor lists were updated.

Overall, I do believe the topic deserves more attention.

There are lots of things that are deserving, but unfortunately, we have only limited workforce available. Most features and contributions in LAMMPS are made by external contributors. The core LAMMPS developers focus more on maintenance and specific tasks they are funded to do.

The topic to hire one of more research software engineers for the expressed purpose to implement requested LAMMPS features comes up repeatedly, but we have not yet agreed on a way to procure the required funding. Basically, there are four main paths: 1. obtain a grant from one or more federal or private funding agencies, 2. donations from users (especially commercial and institutional users), 3. crowdfunding, 4. create or join a non-profit organization created for the purpose of supporting Open Source software development. So far, it has not been obvious which is the best choice and how high the chances are for that choice to be successful.
